drivers/bluetooth/bluetooth.c

```diff
@@ -15,48 +15,73 @@
  */
 
+#include <stdbool.h>
+#include <stddef.h>
 #include "bluetooth.h"
+#include "usb_util.h"
 
-#if defined(BLUETOOTH_BLUEFRUIT_LE)
-#    include "bluefruit_le.h"
-#elif defined(BLUETOOTH_RN42)
-#    include "rn42.h"
-#endif
+/* Each driver needs to define the struct
+ *    const rgb_matrix_driver_t rgb_matrix_driver;
+ * All members (except `task`, `.detect_output` and `system`) must be provided.
+ * Keyboard custom drivers can define this in their own files, it should only
+ * be here if shared between boards.
+ */
 
-void bluetooth_init(void) {
 #if defined(BLUETOOTH_BLUEFRUIT_LE)
-    bluefruit_le_init();
+const bluetooth_driver_t bluetooth_driver = {
+    .init          = bluefruit_le_init,
+    .task          = bluefruit_le_task,
+    .is_connected  = bluefruit_le_is_connected,
+    .send_keyboard = bluefruit_le_send_keyboard,
+    .send_mouse    = bluefruit_le_send_mouse,
+    .send_consumer = bluefruit_le_send_consumer,
+    .send_system   = NULL,
+};
+#    ifdef ENABLE_NKRO
+#        error BlueFruit LE does not support NKRO, do not declare `ENABLE_NKRO`
+#    endif
+
 #elif defined(BLUETOOTH_RN42)
-    rn42_init();
-#endif
-}
+const bluetooth_driver_t bluetooth_driver = {
+    .init          = rn42_init,
+    .task          = NULL,
+    .is_connected  = NULL,
+    .send_keyboard = rn42_send_keyboard,
+    .send_mouse    = rn42_send_mouse,
+    .send_consumer = rn42_send_consumer,
+    .send_system   = NULL,
+};
+#    ifdef ENABLE_NKRO
+#        error RN42 does not support NKRO, do not declare `ENABLE_NKRO`
+#    endif
 
-void bluetooth_task(void) {
-#if defined(BLUETOOTH_BLUEFRUIT_LE)
-    bluefruit_le_task();
 #endif
-}
 
-void bluetooth_send_keyboard(report_keyboard_t *report) {
-#if defined(BLUETOOTH_BLUEFRUIT_LE)
-    bluefruit_le_send_keyboard(report);
-#elif defined(BLUETOOTH_RN42)
-    rn42_send_keyboard(report);
-#endif
+send_output_t desired_send_output = SEND_OUTPUT_DEFAULT;
+
+void set_send_output(send_output_t send_output) {
+    set_send_output_kb(send_output);
+    desired_send_output = send_output;
 }
 
-void bluetooth_send_mouse(report_mouse_t *report) {
-#if defined(BLUETOOTH_BLUEFRUIT_LE)
-    bluefruit_le_send_mouse(report);
-#elif defined(BLUETOOTH_RN42)
-    rn42_send_mouse(report);
-#endif
+__attribute__((weak)) void set_send_output_kb(send_output_t send_output) {
+    set_send_output_user(send_output);
 }
 
-void bluetooth_send_consumer(uint16_t usage) {
-#if defined(BLUETOOTH_BLUEFRUIT_LE)
-    bluefruit_le_send_consumer(usage);
-#elif defined(BLUETOOTH_RN42)
-    rn42_send_consumer(usage);
-#endif
+__attribute__((weak)) void set_send_output_user(send_output_t send_output) {} // do nothing
+
+send_output_t get_send_output(void) {
+    if (desired_send_output == SEND_OUTPUT_AUTO) {
+        // only if USB is **disconnected**
+        if (usb_connected_state()) {
+            return SEND_OUTPUT_USB;
+        }
+        if ((NULL != (*bluetooth_driver.is_connected)) && (bluetooth_driver.is_connected())) {
```

Comment:
I suppose that the logic here should be
```suggestion
        if ((NULL == bluetooth_driver.is_connected) || (bluetooth_driver.is_connected())) {
```
Otherwise `SEND_OUTPUT_AUTO` would be useless if the driver implementation does not provide `is_connected`.

Reply:
I'm assuming `(*func)` is the address of the function ... so I'm testing to see if that address is NULL first, then "short-circuit" operation logic, if that was not NULL then call the actual function.

*(My C may be a bit rusty tho.)*

As to the uselessness ... I agree; and this is why the prior test for `if (usb_connected_state())` assumes that if the USB is not connected, and we cannot test if the BT is connected, then assume nothing is connected.   Also worth remembering that this only gets invoked on a `SEND_OUTPUT_AUTO`.  If the driver is not providing a connection state, then one can always set the `set_send_output()` to `SEND_OUTPUT_BLUETOOTH` or `SEND_OUTPUT_BOTH`.

Reply:
Actually `*` is not needed here, but it ends up being a noop when applied to a pointer to a function:
https://en.cppreference.com/w/c/language/conversion#Function_to_pointer_conversion

Usually `*` is not used with function pointers though, and here it just looks confusing (with a normal pointer `*` would mean that you are accessing data through the pointer, not using the pointer itself as a value).  So probably all cases of `NULL == (*bluetooth_driver.something)` which you have here should have that `*` removed.

Reply:
S@#t ... you're right.  Compiles as the same thing.

![image](https://github.com/qmk/qmk_firmware/assets/366673/2e9683af-239b-4f49-a86e-01564c428a04)

... fixing. 😊